### PR TITLE
Add memset to emscripten_with_builtin_malloc

### LINKED ISSUE #11732: ASan + gmock has a wild pointer error that doesn't happen on other platforms
```
%> git clone https://github.com/trybka/scraps.git
%> cd asan-test
%> emcc -o test.js asan_bug.cc gmock-gtest-all.cc -I. -fsanitize=address -O1 -sTOTAL_MEMORY=1073741824 -g4 --profiling
%> node test.js
```
```
=================================================================
==42==ERROR: AddressSanitizer: unknown-crash on address 0x1000023a at pc 0x00065379 bp 0x0298ff70 sp 0x0298ff70
WRITE of size 2097152 at 0x1000023a thread T0
    #0 0x65379 in __asan_memset+0x65379 (asan-test/test.wasm+0x65379)
    #1 0x8000126b in syscallMmap2 (asan-test/test.js:4715:3
    #2 0x80001282  (JavaScript+0x1282)
    #3 0x80001259 in _emscripten_with_builtin_malloc (asan-test/test.js:4697:10
    #4 0x80001281 in _emscripten_builtin_mmap2 (asan-test/test.js:4737:9
    #5 0x7cd1e in __sanitizer::internal_mmap(void*, unsigned long, int, int, int, unsigned long)+0x7cd1e (asan-test/test.wasm+0x7cd1e)
    #6 0x797c6 in __sanitizer::MmapOrDie(unsigned long, char const*, bool)+0x797c6 (asan-test/test.wasm+0x797c6)
    #7 0x6bb72 in __asan::AppendToErrorMessageBuffer(char const*)+0x6bb72 (asan-test/test.wasm+0x6bb72)

Address 0x1000023a is a wild pointer.
SUMMARY: AddressSanitizer: unknown-crash (asan-test/test.wasm+0x65378) in __asan_memset+0x65378
Shadow bytes around the buggy address:
  0x01fffff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==42==ABORTING
```

### PR DESCRIPTION
**Pull request by tlively**
This was previously causing crashes due to ASan-instrumented memset
being used beneath the ASan boundary in the implementation of
mmap. Unfortunately, these crashes only happened when the allocation
sequence set the heap up in such a manner that the
previously-instrumented memset would intersect a poison region on the
heap, so we don't have a reproducer small enough to make a good test.

As a drive-by, also remove the use of __sys_pipe from libasan so that
it can link successfully with NO_FILESYSTEM.

Fixes #11732.

### CODE REVIEW

**Comment by tlively** [comment]
@sbc100 I ended up backing out the renaming change because this function is called from a few C++ files via EM_ASM, and for some reason that did not like the $camelCase naming convention.

File: tests/test_other.py [hunk 1]
```diff
@@ -9811,7 +9811,7 @@ def test_mmap_and_munmap(self):
     self.do_other_test('mmap_and_munmap', emcc_args)
 
   def test_mmap_and_munmap_anonymous(self):
-    self.do_other_test('mmap_and_munmap_anonymous', emcc_args=['-s', 'NO_FILESYSTEM'])
+    self.do_other_test('mmap_and_munmap_anonymous', emcc_args=['-s', 'NO_FILESYSTEM', '-fsanitize=address', '-s', 'ALLOW_MEMORY_GROWTH=1'])
```
Comment (kripken) [inline tests/test_other.py:9814]: how about adding an asan variation of this test (using parameterize)? there probably isn't any coverage lost by adding asan, but the current way makes it unclear if the test makes sense without asan too. so maybe a comment at least?

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp [hunk 2]
```diff
@@ -277,6 +277,11 @@ bool SignalContext::IsStackOverflow() const {
 #endif  // SANITIZER_GO
 
 bool IsAccessibleMemoryRange(uptr beg, uptr size) {
+#ifdef __wasm__
+  // Avoid pulling in __sys_pipe for the trick below, which doesn't work on
```
Comment (kripken) [inline system/lib/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:281]: I don't see any `__wasm__` checks before this - but lots of ifdefs on `SANITIZER_EMSCRIPTEN`. Maybe that's what we should use?

**Comment by sbc100** [comment]
lgtm % existing comments and test failures.

**Comment by tlively** [comment]
I want to check that this still LGTY wil the new change to make the inclusion of `emscripten_builtin_memset` dependent on ASan being used.

File: emcc.py [hunk 3]
```diff
@@ -1935,9 +1935,16 @@ def include_and_export(name):
...
         shared.Settings.USE_OFFSET_CONVERTER = 1
-        shared.Settings.EXPORTED_FUNCTIONS += ['_memalign', '_emscripten_builtin_memalign',
-                                               '_emscripten_builtin_malloc', '_emscripten_builtin_free',
-                                               '___data_end', '___heap_base', '___global_base']
+        shared.Settings.EXPORTED_FUNCTIONS += [
+            '_memalign',
+            '_emscripten_builtin_memalign',
+            '_emscripten_builtin_malloc',
+            '_emscripten_builtin_free',
+            '_emscripten_builtin_memset',
```
Comment (trybka) [inline emcc.py:1943]: This needs to be guarded by USE_ASAN below because it is only exported for address sanitizer.

File: src/library.js [hunk 4]
```diff
@@ -4511,12 +4515,19 @@ LibraryManager.library = {
...
+    _memset = _emscripten_builtin_memset;
+#endif
     try {
       return func();
     } finally {
       _malloc = prev_malloc;
       _memalign = prev_memalign;
       _free = prev_free;
+#if USE_ASAN
+      _memset = _emscripten_builtin_memset
```
Comment (trybka) [inline src/library.js:4529]: here, should this be _memset = prev_memset ?

Comment (tlively) [reply src/library.js:4529]: Oops, brain fart. Thanks for the catch!

**Comment by trybka** [comment]
Anything blocking this from going in? I've been manually patching this in to run ASan.

**Comment by sbc100** [comment]
No, the fastcomp test failure is a flake.  @tlively feel free to squash and land as-is

**Comment by tlively** [comment]
Oops, I forgot about this. Thanks for the ping!

**Comment by tlively** [comment]
Looks like I was making some requested review changes and fell down a rabbit hole. I'll clean this up and get it landed.

**Pull request merged**